### PR TITLE
Add NCCL Broadcast operation

### PR DESCRIPTION
**Pull request by nvcastet**
Horovod needs to be compiled with HOROVOD_GPU_BROADCAST=NCCL.
Implements #1521.

### CODE REVIEW

**Comment by jessebenson** [comment]
Good to see more NCCL ops!

**Comment by tgaddair** [comment]
Hey @nvcastet, looks like some unit tests are failing:

```
ERROR:tensorflow:2 root error(s) found.
(0) Unknown: Type uint8 is not supported in NCCL mode.
[[node HorovodBroadcast_Cast_0 (defined at /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py:1748) ]]
[[All/_7]]
(1) Unknown: Type uint8 is not supported in NCCL mode.
[[node HorovodBroadcast_Cast_0 (defined at /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py:1748) ]]
0 successful operations.
0 derived errors ignored.
```

**Comment by nvcastet** [comment]
@tgaddair It is finally ready for review :)

File: horovod/common/ops/nccl_operations.h [hunk 1]
```diff
@@ -38,15 +38,12 @@ struct NCCLContext {
...
+class NCCLOp {
 public:
-  NCCLAllreduce(NCCLContext* nccl_context, CUDAContext* cuda_context,
-                HorovodGlobalState* global_state)
-      : CUDAAllreduce(cuda_context, global_state),
-        nccl_context_(nccl_context){};
-
-  Status Execute(std::vector<TensorTableEntry>& entries,
-                 const Response& response) override;
+  NCCLOp(NCCLContext* nccl_context, HorovodGlobalState* global_state)
```
Comment (tgaddair) [inline horovod/common/ops/nccl_operations.h]: In a previous PR (https://github.com/horovod/horovod/pull/1480/files#diff-5798de9e6fdc83e7f7bf0fa8344a56c0R65) we opted to use delegation to solve this problem by introducing `CUDAOpContext`.  It looks like we could do something similar here, which would make things in our codebase more consistent.  What do you think?

File: horovod/common/wire/message.fbs [hunk 2]
```diff
@@ -28,7 +28,8 @@ enum DataType:byte {
     HOROVOD_FLOAT16 = 6,
     HOROVOD_FLOAT32 = 7,
     HOROVOD_FLOAT64 = 8,
-    HOROVOD_BOOL = 9
+    HOROVOD_BOOL = 9,
+    HOROVOD_BYTE = 10
```
Comment (tgaddair) [inline horovod/common/wire/message.fbs]: I'm a little confused by `HOROVOD_BYTE`.  In what cases would it be used in place of `HOROVOD_UINT8`?

File: horovod/common/ops/nccl_operations.h (hunk 1 above)
Comment (nvcastet) [reply horovod/common/ops/nccl_operations.h]: I agree to try to keep things consistent through the code base. Let me give it a try.

File: horovod/common/wire/message.fbs (hunk 2 above)
Comment (nvcastet) [reply horovod/common/wire/message.fbs]: I was under the impression that we needed to keep the enums consistent between `enum DataType` in [message.h](https://github.com/horovod/horovod/blob/master/horovod/common/message.h) and `enum DataType` in [message.fbs](https://github.com/horovod/horovod/blob/master/horovod/common/wire/message.fbs)/[message_generated.h](https://github.com/horovod/horovod/blob/master/horovod/common/wire/message_generated.h). Am i incorrect? `HOROVOD_BYTE` was already in [message.h](https://github.com/horovod/horovod/blob/master/horovod/common/message.h).

Comment (tgaddair) [reply horovod/common/wire/message.fbs]: Ah, I see what happened.  The `HOROVOD_BYTE` dtype was added in https://github.com/horovod/horovod/commit/aa605d6de1e07d1184e8793b6d4f65cee9d4efb3 to abstract an API call to `MPI_Bcast`, but was subsequently simplified in https://github.com/horovod/horovod/commit/669eb9e0dde25f7475e6a31630ae4b9a704ae106.

The result is that `HOROVOD_BYTE` is no longer used for anything, so I think we can safely remove it.

**Pull request merged**